test/program_test.go

```diff
@@ -0,0 +1,242 @@
+		})
+
+		It("should fail if given a non-existent program.", func() {
+			stack := pulumiv1.Stack{
+				Spec: shared.StackSpec{
+					Stack:   randString(),
+					Backend: fmt.Sprintf("file://%s", backendDir),
+					ProgramRef: &shared.ProgramReference{
+						Name: randString(),
+					},
+					EnvRefs: map[string]shared.ResourceRef{
+						"PULUMI_CONFIG_PASSPHRASE": shared.NewLiteralResourceRef("password"),
+					},
+				},
+			}
+			stack.Name = randString()
+			stack.Namespace = "default"
+
+			Expect(k8sClient.Create(context.TODO(), &stack)).To(Succeed())
+
+			// wait until the controller has seen the stack object and completed processing it
+			var s pulumiv1.Stack
+			Eventually(func() bool {
+				err := k8sClient.Get(context.TODO(), types.NamespacedName{Namespace: stack.Namespace, Name: stack.Name}, &s)
+				if err != nil {
+					return false
+				}
+				if s.Generation == 0 {
+					return false
+				}
+				return s.Status.ObservedGeneration == s.Generation
+			}, "20s", "1s").Should(BeTrue())
+
+			Expect(s.Status.LastUpdate.State).To(Equal(shared.FailedStackStateMessage))
+			Expect(apimeta.IsStatusConditionFalse(s.Status.Conditions, pulumiv1.ReadyCondition)).To(BeTrue())
+		})
+
+		It("should fail if given a syntactically correct but invalid program.", func() {
+			prog := programFromFile("./testdata/test-program-invalid.yaml")
+			Expect(k8sClient.Create(context.TODO(), &prog)).To(Succeed())
+
+			stack := pulumiv1.Stack{
+				Spec: shared.StackSpec{
+					Stack:   randString(),
+					Backend: fmt.Sprintf("file://%s", backendDir),
+					ProgramRef: &shared.ProgramReference{
+						Name: prog.Name,
+					},
+					EnvRefs: map[string]shared.ResourceRef{
+						"PULUMI_CONFIG_PASSPHRASE": shared.NewLiteralResourceRef("password"),
+					},
+				},
+			}
+			stack.Name = randString()
+			stack.Namespace = "default"
+
+			Expect(k8sClient.Create(context.TODO(), &stack)).To(Succeed())
+
+			// wait until the controller has seen the stack object and completed processing it
+			var s pulumiv1.Stack
+			Eventually(func() bool {
+				err := k8sClient.Get(context.TODO(), types.NamespacedName{Namespace: stack.Namespace, Name: stack.Name}, &s)
+				if err != nil {
+					return false
+				}
+				if s.Generation == 0 {
+					return false
+				}
+				return s.Status.ObservedGeneration == s.Generation
+			}, "20s", "1s").Should(BeTrue())
+
+			Expect(s.Status.LastUpdate.State).To(Equal(shared.FailedStackStateMessage))
+			Expect(apimeta.IsStatusConditionFalse(s.Status.Conditions, pulumiv1.ReadyCondition)).To(BeTrue())
+		})
+	})
+
+})
+
+func programFromFile(path string) pulumiv1.Program {
+	prog := pulumiv1.Program{}
+	prog.Name = randString()
+	prog.Namespace = "default"
+
+	programFile, err := os.ReadFile(path)
+	if err != nil {
+		fmt.Printf("%+v", err)
+		Fail(fmt.Sprintf("Couldn't read program file: %v", err))
+	}
+	spec := pulumiv1.ProgramSpec{}
+	err = yaml.Unmarshal(programFile, &spec)
+	if err != nil {
```

Comment:
You can use `ExpectWithOffset(1, err).ToNot(HaveOccurred())`, which is a bit more consistent and compact. The `WithOffset` bit refers to an offset in the call stack, so that if you are calling this from test cases, the test report highlight the failing line in the test case rather than the failing line here.